### ISSUE TITLE
How to initialize redux store with data from Gatsby page query?

### ISSUE THREAD

**Issue by Kadrian** [gatsbyjs/gatsby#23239]
## Summary

How to initialize redux store with data from a Gatsby page query?

Is there a way to use the `wrapRootElement` function on ssr?

**Comment by Js-Brecht** [comment]
This is how I handle that in one of the projects I'm working on:

```typescript
import React from 'react';
import { Provider } from 'react-redux';
import { StaticQuery, graphql, GatsbyBrowser } from 'gatsby';
import { createStore } from '~/store/store';
import { actions } from '~/store';
import { MenuDataQuery } from '@/graphql/types';

const { hydrateMenuState } = actions.ui.menu;

export const wrapRootElement: GatsbyBrowser['wrapRootElement'] = ({ element }): React.ReactNode => {
    const store = createStore();
    const dispatch = store.dispatch;
    return (
        <StaticQuery
            query={graphql`
                query MenuData {
                    allMenuItem {
                        edges {
                            node {
                                id
                                url
                                slug
                                label
                                acl
                                routes
                                children {
                                    id
                                }
                            }
                        }
                    }
                }
            `}
            render={(data: MenuDataQuery) => {
                dispatch(hydrateMenuState(data));
                return <Provider store={store}>{element}</Provider>;
            }}
        />
    );
};
```

It doesn't really do a lot at the moment, but the concept is there.

It's also possible to use hooks if you wrap it in another React component

**Comment by Js-Brecht** [comment]
I keep that `wrapRootElement` function in a separate module, because I use it the same way in both `gatsby-browser` and `gatsby-ssr`.

**Comment by Kadrian** [comment]
I see. Thanks for sharing that idea! This would work if I always had the same query for many pages. But I think I'm looking more for the individual page queries.

Yes, I dispatching to the store inside of a useEffect, once the page renders, is what I'm doing right now. But then it's not on the server-side anymore :-/

**Comment by Js-Brecht** [comment]
Why are you doing it inside of `useEffect()`?

**Comment by Kadrian** [comment]
Where else would I dispatch an action from the page?

**Comment by Js-Brecht** [comment]
What kind of action are you trying to dispatch?  What does it do?

**Comment by Kadrian** [comment]
I'm dispatching an action that pretty much "initializes" the store with some data from the page query. And what I'm looking for is a way to do the same on the server side, just like you did above, but with the specific data from the specific page

**Comment by Js-Brecht** [comment]
If the action is synchronous, I don't imagine it being an issue just doing it from the root of the function.

```js
const FooPage = ({ data }) => {
   const dispatch = useDispatch();
   dispatch(someAction(data));

   return <div>Content</div>
}
```

I'd have to do some testing to see how it would work with asynchronous actions.  I don't think SSR will wait if it's asynchronous.

If you want to dispatch the same action on every page, you could use `wrapPageElement`.  It receives all of the same props your pages do, so it has access to `pageContext` and `data`.

**Comment by Kadrian** [comment]
Hmm, in your example I would dispatch on every render. I'd probably want to avoid that, since I'll only need to initialize the redux store once.

You're right, `wrapPageElement` gets the data through the prop. So I could dispatch the action from there. I guess that would really work. 

But it still feels like a detour, when I could simply pass the data one level higher as an `initialState` to `createStore`

**Comment by Js-Brecht** [comment]
If you're trying to do it at the page level, wouldn't that indicate that the data will be different on each page?

**Comment by Kadrian** [comment]
Yes it will be (the data in my case depends on the page url).

**Comment by Js-Brecht** [comment]
If you only passed it as initial state, then that state wouldn't change when people navigate to different pages.  They would be stuck with the state of the page they first landed on.

**Comment by Kadrian** [comment]
That's true, but it's page-specific state. So that part of the state doesn't concern other pages

**Comment by Js-Brecht** [comment]
If it's page-specific state, then each page should take care of changing the state based on its context.

Your `createStore` (should) be getting called for every page during SSR.  That way, each page is rendered with its own state... BUT, when you're in the browser, it's only called once, and that's with the state of whatever page is landed on first.  In order to change that state, you would need to `dispatch()` an action to change it (from below `wrapRootElement`).

Makes sense to me to only initialize the store with the global state, and on every page change `dispatch()` a new action to update the state for that current page.

Perhaps I'm misunderstanding 😕

**Comment by Kadrian** [comment]
Hmm I think you're understanding correctly. I guess you're right.
So maybe what I can really do is use a single global wrapper like in your very first example to initialize the store at least with all info that comes from the URL. Then if that information is available, I think my component can simply use that on the first render, so already on the server side, to do what's necessary without having to do the initial-dispatch-workaround.

**Comment by Js-Brecht** [comment]
Sure.  My initial example could be refactored to this:
```js
export const wrapRootElement: GatsbyBrowser['wrapRootElement'] = ({ element }): React.ReactNode => {
    return (
        <StaticQuery
            query={graphql`
                query MenuData {
                    allMenuItem {
                        edges {
                            node {
                                id
                                url
                                slug
                                label
                                acl
                                routes
                                children {
                                    id
                                }
                            }
                        }
                    }
                }
            `}
            render={(data: MenuDataQuery) => {
                return <Provider store={createStore(data)}>{element}</Provider>;
            }}
        />
    );
};
```

I just intended to remove `createStore()` from that function altogether.  `gatsby-ssr` would wrap my `wrapRootElement`, and call `createStore()` itself.  `gatsby-browser` would just use a singleton

**Comment by Js-Brecht** [comment]
Did that answer your questions, @Kadrian?

**Comment by Kadrian** [comment]
Yes, thanks!

**Comment by tranlehaiquan** [comment]
I try the same, but I got this error when build:

wrap-with-provider.tsx

```typescript
import React from 'react';
import { Provider } from 'react-redux';
import { graphql, StaticQuery } from 'gatsby';

import createStore from './src/store/index';

// eslint-disable-next-line react/display-name,react/prop-types
export default ({ element }) => {
  return (
    <StaticQuery
      query={graphql`
        query {
          site {
            siteMetadata {
              title
              description
              author
            }
          }
        }
      `}
      render={(data: any) => {
        console.log(data);
        const store = createStore();
        // store.dispatch(setStays(data.allStaysJson.nodes));
        return <Provider store={store}>{element}</Provider>;
      }}
    />
  );
};

```


```shell
info bootstrap finished - 5.208s
success run page queries - 0.018s - 5/5 283.01/s
success write out requires - 0.004s
success Building production JavaScript and CSS bundles - 5.541s
success Rewriting compilation hashes - 0.017s
error Generating SSR bundle failed

Can't resolve 'public/page-data/sq/d/63159454.json' in '/Users/myname/GitHub/devchallenges/windbnb'

If you're trying to use a package make sure that 'public/page-data/sq/d/63159454.json' is installed. If you're trying to use a local file make sure that the path is correct.
```

My public/page-data/sq/d/63159454.json

```Json
{"data":{"site":{"siteMetadata":{"title":"Gatsby Default Starter","description":"Kick off your next, great Gatsby project with this default starter. This barebones starter ships with the main Gatsby configuration files you might need.","author":"@gatsbyjs"}}}}
```

**Comment by Js-Brecht** [comment]
@tranlehaiquan have you tried a `gatsby clean`?

**Comment by tranlehaiquan** [comment]
@Js-Brecht Yes, I have just tried it and it not helping.

**Comment by Js-Brecht** [comment]
Hmm... it’s hard to tell what might be happening, but I can tell you it’s not a redux thing.  It should probably be a new ticket

**Comment by yidongw** [comment]
@Js-Brecht what should I do without graphQL?

**Comment by Js-Brecht** [comment]
@yidongw what do you mean?

**Comment by yidongw** [comment]
@Js-Brecht 
I'm using mobx-state-tree for storing data, and it contains some product data that doesn't change very often, so I want to get those data ready in the build time. Fetching those data is written in mobx-state-tree action, and it is REST API instead of graphQL.

I implemented wrapRootElement in gatsby-ssr.js and gatsby-browser.js. wrapRootElement is the same function that creates a store and wrapping the element with context provider. According to some tutorial.
```
import { MainContext, MainModel } from "../stores/MainStore"

export default ({ element }) => {
  const store = MainModel.createStore({})
  return <MainContext.Provider value={store}>{element}</MainContext.Provider>
}
```
I also need to createPage in gatsby-node.js according to [using-gatsby-without-graphql](https://www.gatsbyjs.com/docs/how-to/querying-data/using-gatsby-without-graphql/), but all the data fetching is currently done in the store, so in order to fetch the data, I need to somehow access the store that was created in the gatsby-ssr.js or gatsby-browser.js.

How am I supposed to do that? Or I should use a different approach?

Your answer is the closest approach, but you are using graphQL, so I don't know what I should do with my stuff.

I also posted my questions [here](https://github.com/gatsbyjs/gatsby/discussions/31197).
